File: Trees/parse_input.cpp
#include <queue>
#include "parse_input.h"


void checkNode(const TreeNode* cur_node, const std::string& n_val){
    if (!cur_node) throw std::runtime_error("The provided list is incorrect, no parent node for " + n_val);
}


TreeNode* string2tree(const std::string &line){
    std::istringstream ss(line);
    std::string n_val;

    // create root
    ss >> n_val;
    TreeNode* root;
    if (!n_val.empty() && n_val != "null"){
        root = new TreeNode(std::stoi(n_val));
    }
    else return nullptr;

    // fill in level by level in order: left -> right, left -> right
    std::queue<TreeNode*> level;
    level.push(root);
    bool left = true;
    TreeNode* cur_node;
    while (ss.rdbuf()->in_avail()){
        if (left){
            if (level.size()) {
                cur_node = level.front();
                level.pop();
            }
            else cur_node = nullptr;
        }
            
        ss >> n_val;
        if (n_val != "null") {
            checkNode(cur_node, n_val);
            TreeNode* new_node = new TreeNode(std::stoi(n_val));
            if (left)
                cur_node->left = new_node;
            else
                cur_node->right = new_node;
            level.push(new_node);
        }

        left = !left;
    }
    return root;
}

File: Trees/parse_input.py
from queue import Queue


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def check_node(node, n_val):
    if not node:
        raise RuntimeError('The provided list is incorrect, no parent node for ' + n_val)
    

def string2tree(line):
    node_vals = Queue()
    for v in line.split(' '):
        node_vals.put(v)

    n_val = node_vals.get() if not node_vals.empty() else 'null'
    if n_val and n_val != 'null':
        root = TreeNode(int(n_val))
    else:
        return None

    # fill in level by level in order: left -> right, left -> right
    level = Queue()
    level.put(root)
    left = True
    while not node_vals.empty():
        if left:
            if not level.empty():
                cur_node = level.get()
            else:
                cur_node = None

        n_val = node_vals.get()
        if n_val != 'null':
            check_node(cur_node, n_val)
            new_node = TreeNode(int(n_val))
            if left:
                cur_node.left = new_node
            else:
                cur_node.right = new_node
            level.put(new_node)

        left = not left
    
    return root


__all__ = ['TreeNode', 'string2tree']